### ISSUE TITLE
Database agnostic where ILIKE 

### ISSUE THREAD

**Issue by esatterwhite** [knex/knex#2265]
As mentioned in Issue #233 @ErisDS  there are no way to do case insensitive string comparison (which would work the same way that ilike in postgres). Would be nice to have one.

EDIT: made description to be about just one single feature (original issue had a lot more discussion around the topic)

**Comment by esatterwhite** [comment]
Generally speaking there is no consistent way to specify the case sensitivity of your queries through knex. 
for example w/ mysql it is case insensitive. and doing `where('a' 'ilike' 'b')` doesn't work, you have to send something like

```javascript
whereRaw(util.format("%s LIKE BINARY '%s'",field, '%' + term + '%'));
```
swapping DB engines at this point is not possible. The query parser need to better handle what `ilike` means.

perhaps just a function ? 

```javascript
knex.ilike(<field>, <value>)
```

**Comment by elhigu** [comment]
I removed copy-paste from original issue and changed the title, because it just would have made this issue to be as unclear about its purpose that the original was.

Since this seems to be just about having `.where` variant for case insensitive query filterin, then the API should be the same way with other `.where` methods. 

`.whereILike(...)` and `.orWhereILike(...)` (probably there is no need for ends with / starts with variants, because they should trivial to implement with ilike and %).

**Comment by esatterwhite** [comment]
Yes, the additional methods aren't strictly needed, I'm pretty sure the wild card character is the same across the board. 

You will need `whereLike` and `orWhereLike` variants though.

**Comment by elhigu** [comment]
Those are kind of different issue not related to case sensitivity, but it will be ok to add them at the same time if one wants to have them as special methods. I've been always just using .where(col, 'like', something) since there is no compatibility problems with that between dbs (well actually I always target postgres only, but anyways I would use it that way even if I had to support multiple dbs).

**Comment by esatterwhite** [comment]
Sent from my BlackBerry 10.                                                                                                                                                                                                                From: Mikael LepistöSent: Wednesday, October 11, 2017 3:46 PMTo: tgriesser/knexReply To: tgriesser/knexCc: esatterwhite; AuthorSubject: Re: [tgriesser/knex] Database agnostic where ILIKE (#2265)Those are kind of different issue not related to case sensitivity, but it will be ok to add them at the same time if one wants to have them as special methods. I've been always just using .where(col, 'like', something) since there is no compatibility problems with that between dbs.

—You are receiving this because you authored the thread.Reply to this email directly, view it on GitHub, or mute the thread.


  
  




{"api_version":"1.0","publisher":{"api_key":"05dde50f1d1a384dd78767c55493e4bb","name":"GitHub"},"entity":{"external_key":"github/tgriesser/knex","title":"tgriesser/knex","subtitle":"GitHub repository","main_image_url":"https://cloud.githubusercontent.com/assets/143418/17495839/a5054eac-5d88-11e6-95fc-7290892c7bb5.png","avatar_image_url":"https://cloud.githubusercontent.com/assets/143418/15842166/7c72db34-2c0b-11e6-9aed-b52498112777.png","action":{"name":"Open in GitHub","url":"https://github.com/tgriesser/knex"}},"updates":{"snippets":[{"icon":"PERSON","message":"@elhigu in #2265: Those are kind of different issue not related to case sensitivity, but it will be ok to add them at the same time if one wants to have them as special methods. I've been always just using .where(col, 'like', something) since there is no compatibility problems with that between dbs."}],"action":{"name":"View Issue","url":"https://github.com/tgriesser/knex/issues/2265#issuecomment-335942845"}}}

**Comment by esatterwhite** [comment]
There are compatibility problems, that is the issue. 

```sql
WHERE foo LIKE %BAR%
```
This will return different results between engines.

*mysql* will match `foobar`, `fooBaR`, `fooBAR`, `bar`
*postgres* will only match `fooBAR`

The syntax may be the same for most engines, but the behavior is not. Mysql, for example, doesn't support `ILIKE`. But the default behavior for `LIKE` behaves in a case insensitive manner. 

so if you only add `whereILike` and use `where('field', 'like', 'value')` in mysql, they will do the exact same thing. a case **insensitive** match

Additionally, if you were to use the same code, and switch to PG for some reason you'd start getting back different results for the same data sets.

**Comment by elhigu** [comment]
@esatterwhite ah right, I didn't knew/remember that even `LIKE` operator does not work the same in every dialect. Yeah makes sense to have `.whereLike` variants too.

**Comment by ShinRai1090** [comment]
sorry for necro~ing on this, but has this resulted in any function / variation that can be used? I need to perform a case insensitive compare in postgres,

**Comment by kibertoad** [comment]
Not yet. PR would be appreciated!

**Comment by elhigu** [comment]
@ShinRai1090 did you try `.where('field', 'ilike', 'value')`?

**Comment by kibertoad** [comment]
Released in 1.0.1.

**Comment by OlivierCavadenti** [comment]
I reopen this issue for the orWhereLike and orWhereILike implementation.

**Comment by kibertoad** [comment]
Released in 1.0.4
